### PR TITLE
router ssl squash

### PR DESCRIPTION
**Pull request by pweil-**
This is a pre-final PR for comments and clean up.  This will also lead to a follow up doc PR to migrate routing.md into the documentation project

### CODE REVIEW

**Comment by pweil-** [comment]
For your review:

1.  Implementation of adding certificates to a route for the following scenarios:
    1. edge termination: router accepts tcp traffic with SNI on :443.  Use a cert, key, and ca cert at the router.  Send decrypted http traffic to endpoints
    1. pod termination: router accepts tcp traffic with SNI on :443.  Router sends encrypted tcp traffic to  endpoints
    1. reencrypt termination: router accepts tcp traffic with SNI on :443.  Use a cert, key, and ca cert at the router.  Create a tcp ssl backend that does a health check with a ca cert on the endpoints
    1. non-sni: router tries to use a generic cert to decrypt traffic.  Then attempts to check for a reencrypt backend, if none is found then use a regular http backend
1.  refactor of route domain to handle certificates.  Summary: `ServiceUnit` is 1:1 for a service.  `ServiceUnit.EndpointTable` still contains all known endpoints for the service.  `ServiceUnit.ServiceAliasConfig` is 1:1 for a route.  Certificate information is stored on the `ServiceUnit.ServiceAliasConfig`
1. Validations and validation test cases to support the constructs above
1. Refactor of existing tests to support the new domain
1. HAProxy implementation of the template router to support the scenarios above.  When reviewing the template please keep this diagram in mind: https://docs.google.com/a/redhat.com/drawings/d/1fRsvKg1iuY1EfQ-C3O6s8lxxouapzuCwVz7jFUKDUsc/edit

Open Issues:

1.  HAProxy prompts for a password when using an encrypted private key and does not provide a way to automate the response.  Currently we need a non passphrase protected private key
1.  Private key passphrase is stored in clear text.  See above, currently validation prevents users from storing a passphrase as an temporary solution
1.  Loading of the files from the cli is done in the REST implementation @ffranz - let me know the best way to handle this and I'll make the correct updates
1.  `certmanager.go` is not writing the ca cert - didn't get to it yet, will address Monday
1.  Bug: deleting a route removes all service alias configs for the service name
1.  Bug: route mods don't appear to be working as intended (not due to new code, but will address)


I have tested the following scenarios so far:

1.  plain route (non-ssl) 
1.  edge term only
1.  pod term only
1.  edge term and pod term in combination (2 services, 2 routes)

@pmorie @rajatchopra @smarterclayton

**Comment by pweil-** [comment]
@smarterclayton : per our discussions in IRC on Friday, next step in the router lifecycle needs to be determining how much of this can be EOL'd with all the recent PRs in upstream.

**Comment by pweil-** [comment]
Just FYI, I will squash the fixes I'm committing, I just have to go through the commit/push/build process for the images for a full test.

**Comment by smarterclayton** [comment]
@thockin as per the discussion at the Kube face to face, a lot of these capabilities are useful for abstract ingress of traffic.  i think it's generally useful enough on its own to be a Kube plugin like logging (even without the rest resource)

File: docs/routing.md [hunk 1]
```diff
@@ -144,7 +144,36 @@ route.json
...
-    
+
+## Securing Your Routes
+
+Creating a secure route to your pods can be accomplished by specifying the TLS Termination of the route and, optionally,
+providing certificates to use.  TLS Termination falls in the following configuration buckets:
+
+#### Edge Termination
+Edge termination means that TLS Termination occurs prior to traffic reaching the pod.  TLS certificates are served by either 
+inspecting the SNI host sent in the request and finding a certificate for that host (matching the `CommonName` in the
```
Comment (smarterclayton) [inline docs/routing.md]: Shouldn't this fall through to generic certs regardless of whether it's sni?

File: docs/routing.md [hunk 2]
```diff
@@ -144,7 +144,36 @@ route.json
...
+certificate file) or, for non-SNI requests by serving a generic certificate.
+
+Edge termination is configured by setting `TLS.Termination` to `edge` on your `route` and by specifying the `CertificateFile`
+and `KeyFile` (at a minimum).  
+
+#### Pod Termination
+Pod termination is a passthrough mechanism to send encrypted traffic straight to your pod.  Pod termination relies on 
+SNI hosts to determine where to route the traffic.  
+
+Pod termination is configured by setting `TLS.Termination` to `pod` on your `route`.  No other information is required.
```
Comment (smarterclayton) [inline docs/routing.md]: I don't think we want to couple routes to pods - instead this should just refer to "backend" or whatever the term you're using is.

File: docs/routing.md (hunk 1 above)
Comment (pweil-) [inline docs/routing.md]: I don't follow, if we always use a generic certificate then we aren't really offering custom ssl edge termination with a user's certificate.  Can you describe in more detail what you're thinking?  This is the flow that I'm currently trying to achieve: https://docs.google.com/a/redhat.com/drawings/d/1fRsvKg1iuY1EfQ-C3O6s8lxxouapzuCwVz7jFUKDUsc/edit

Full disclosure: I'm getting intermittent failures from haproxy today in my edge term testing that I have yet to explain so I'm not opposed to a pivot here.

File: docs/routing.md (hunk 2 above)
Comment (pweil-) [inline docs/routing.md]: I can rename this to passthrough termination so it doesn't directly reference the term 'pod'.  The coupling is in name only, the functionality is that the router will create a tcp backend, accept a connection over :443, and route it by checking the SNI host to the tcp backend

File: docs/routing.md (hunk 1 above)
Comment (smarterclayton) [inline docs/routing.md]: Using SNI is orthogonal to providing a cert.  If I'm using SNI, the cert I use could either by my cert or the generic cert.  If I'm not using SNI, the cert I use could either be my cert or the generic cert.  The last part of the sentence didn't make sense.


----- Original Message -----

**Comment by pweil-** [comment]
Rebased, closing this for final testing and PR

**Pull request closed without merging**